### PR TITLE
Upsert by query

### PR DESCRIPTION
**Pull request by seut**

### CODE REVIEW

File: sql/src/test/java/io/crate/analyze/InsertFromSubQueryAnalyzerTest.java [hunk 1]
```diff
@@ -197,8 +203,50 @@ public void testImplicitTypeCasting() throws Exception {
...
+        Assert.assertThat(statement.onDuplicateKeyAssignments().size(), is(1));
+
+        for (Map.Entry<Reference, Symbol> entry : statement.onDuplicateKeyAssignments().entrySet()) {
+            assertThat(entry.getKey(), isReference("name"));
+            assertThat(entry.getValue(), isLiteral("Arthur", StringType.INSTANCE));
+        }
+    }
+
+    @Test
+    public void testFromQueryWithOnDuplicateKeyValues() throws Exception {
```
Comment (mfelsche) [inline sql/src/test/java/io/crate/analyze/InsertFromSubQueryAnalyzerTest.java]: what about a failing test with values(unknown_column) ?

File: sql/src/main/java/io/crate/planner/projection/SourceIndexWriterProjection.java [hunk 2]
```diff
@@ -75,29 +79,41 @@ public SourceIndexWriterProjection(String tableName,
...
+        for (int i = 0; i < primaryKeys.size(); i++) {
+            idSymbols.add(new InputColumn(i, null));
+        }
+
+        partitionedBySymbols = new ArrayList<>(partitionedBy.size());
         for (int i = 0, length = partitionedBy.size(); i < length; i++) {
             int idx = primaryKeys.indexOf(partitionedBy.get(i));
+            Symbol partitionSymbol;
+            if (partitionValues.size() > i) {
+                // copy from into partition, do NOT define partition symbols
```
Comment (mfelsche) [inline sql/src/main/java/io/crate/planner/projection/SourceIndexWriterProjection.java]: no tests verifying this behaviour?

**Comment by mfelsche** [comment]
otherwise ack

File: sql/src/main/java/io/crate/planner/projection/SourceIndexWriterProjection.java (hunk 2 above)
Comment (seut) [inline sql/src/main/java/io/crate/planner/projection/SourceIndexWriterProjection.java]: afaik there are tests for this, this behaviour is not new

File: sql/src/test/java/io/crate/analyze/InsertFromSubQueryAnalyzerTest.java (hunk 1 above)
Comment (seut) [inline sql/src/test/java/io/crate/analyze/InsertFromSubQueryAnalyzerTest.java]: see regression tests at insert integration test

**Pull request merged**